### ISSUE TITLE
Apostrophe makes too many queries

### ISSUE THREAD

**Issue by boutell** [apostrophecms/apostrophe#651]
This is one of those deep ongoing non-emergency thought piece issues.

Apostrophe groups together all the widgets on the page (or index of pieces...) by type and calls load for each type, which is pretty good and not too weird to grasp. For pieces based widgets the next step is an efficient single query, if things were being selected by id.

Then those pieces call their own widget loaders, and again one query occurs per widget type as long as they are pieces widgets.

Any non pieces based widget blows this pattern away though. Loaders run one per widget. This could be the issue with UE, I need to study it more closely.

Basically, figure out why UE makes so many _separate_ queries for images pieces on its homepage. It really hurts performance.

A cool idea which may make sense after I decipher that:

`req.cork()` (before we start calling loaders)

Calls for this req that would normally cause cursors to proceed to do real mongo work now just pile them up in a queue.

`req.uncork()` (when we run out of loaders to call so far)

Apostrophe dices up the cursors looking for cases it can merge because they have compatible enough, simple enough criteria that it can make one big query and then dish out the objects to the appropriate callbacks, reimplementing criteria and projection for simple cases, punting others to run normally as separate queries.

`req.uncork()` automatically recorks just before it starts firing callbacks, so we can queue up another batch of queries. This continues until we run out of queries in the queue.

So in the ideal case we get one query per level of join / level of widget loaders, which is an `O(log N)` sort of thing.

There was a mechanism like this in 0.5 specifically for slideshow widgets which was generalized later, but it had weird semantics because your widget loader callback would fire without actually having loaded yet. In this scenario, your widget's load callback only fires when you've truly loaded. It's the model layer that is made clever and magical, hopefully in a backwards compatible way.

cc: @agilbert @jth- @colpanik @kyjoya @mcoppola anybody who feels like having their brain hurt like mine does thinking about it, but in a cool way

**Comment by boutell** [comment]
I hacked in some logging to see all the Apostrophe cursors just before they are converted to mongo queries.

The home page of UE made 46 queries. That's a lot.

TL;DR, here's what I learned and where I need help:

* **Our default behavior for fetching ancestors and descendants is lousy and I can't decide if I get to call it a bug and fix it, please help me.** We already set `areas: false`, I think this is a reasonable precedent and we can set `joins: false` too; a default projection would be even better but it may be too late not to call that a bc break. Your thoughts? What behavior do you expect?
* **The home page getting fetched twice sounds like a possible Apostrophe bug**
* **Fetching all the pieces-pages for a piece type without a projection sounds like a bug,** this is the code for generating URLs for pieces, I believe it is at least trying to do a projection and may simply be buggy. I will treat it as a bug and fix it as such
* **UE has some issues unique to it,** like loading every news item and every project without projection on the home page, that are probably project specific and can be fixed there and don't really shed light on Apostrophe.
* **There are a lot of queries with compatible criteria and projections that could be magically merged by really clever code** without making A2 development harder or weirder, as proposed in this issue. I think we'll have trouble doing another site at the complexity level of FM's pages without it. So that idea lives on and big thoughts about how to implement it will continue.

### ALL THE QUERIES AND WHAT THEY DO

* Fetch `global`
* Fetch the following pages, with a tight projection: news, projects, excellence blog. I assume: it's a nav widget that lives in `global`.
* Fetch a longer list of pages, similar projection. Includes all of their locations and a few other pages. I assume another nav widget in `global`.
* Fetch a handpicked subset of the services pages (probably all of them in a particular order). Another widget on `global`. Another tight projection.
* Fetch all the markets. NO PROJECTION.
* Fetch four specific pages, with a tight projection. This is a widget in the footer.
* Fetch the excellence blog page, with a tight projection. A second join in the footer widget.
* Fetch terms of use, privacy policy, with a tight projection. Yet another join in the footer widget.
* Fetch an image used as footer background.
* Fetch the homepage, with ancestors and children and children of ancestors. Joins are turned on (because it is the current page).
* Fetch our history, excellence blog, faces of urban with a tight projection. A widget on the homepage, most likely.
* Grab five apostrophe-images with one query. These are slides for the homepage marquee.
* Grab featured projects. The projection is a little generous because we need to connect them with markets and services, but it is well-considered.
* Grab the markets related to those projects. Tight projection.
* Grab the services related to those projects. Tight projection.
* Grab all the projects-pages. NO PROJECTION. This might be the "find best page for piece" code for projects.
* Grab four images which illustrate the featured projects.
* Grab ALL PUBLISHED newsItems. NO LIMIT. Projection is sensible, includes thumbnail.
* Grab all news pages. NO PROJECTION. This might be the "find best page for piece" code for news items.
* Grab an UNHOLY SHIT-TON OF IMAGES in one query, by id. These appear to be all of the illustrations for all of the news items that have ever existed (see above).
* Grab the 2 most recent news items tagged news with 1 query and a sensible projection that includes a thumbnail.
* Grab the 2 thumbnails that illustrate those news items.
* Grab ALL PUBLISHED BLOG POSTS. NO LIMIT.  Projection is sensible, includes thumbnail.
* Grab all news pages, presumably to set URLs for the above. NO PROJECTION.
* Grab ALL IMAGES associated with published blog posts.
* Grab ancestors of home page (EMPTY $IN ARRAY, COULD BE SKIPPED). Areas are off but JOINS ARE ON, for a page with ancestors that are custom-pages this could do a lot of work. NO PROJECTION. **We should have set a default projection for ancestors. That's a bc break now** **We should have set a default projection for descendants. That's a bc break now** but we can standardize on it ourselves and in sites generated by the apostrophe CLI.
* Grab descendants of home page. Areas are off, but JOINS ARE ON,  NO PROJECTION. For a page with descendants that use joins in custom-pages schemas, this could do a lot of work. **We should have set a default projection for descendants. That's a bc break now** but we can standardize on it ourselves and in sites generated by the apostrophe CLI.
* Grab tabs. Areas are off, but JOINS ARE ON,  NO PROJECTION. For a site with tabs that use joins in custom-pages schemas, this could do a lot of work. **We should have set a default projection for descendants. That's a bc break now** but we can standardize on it ourselves and in sites generated by the apostrophe CLI.
* Grab all markets. Tight projection.
* Grab all services. Tight projection.
* Grab the **home page, again**, with its children, but not its ancestors (if there were any). ?????
* Grab "Our History," "Excellence Blog," "Faces of Urban" again...
* **DO EVERYTHING FROM "FETCH THE HOME PAGE" AGAIN** (another 20 or so wasted queries).

**Comment by boutell** [comment]
I assigned this ticket to everyone because it's tracking research and ideas into Apostrophe's performance, and includes some questions I need help on. Here's the big current question I need your help on:

**Our default behavior for fetching ancestors and descendants is lousy and I can't decide if I get to call it a bug and fix it, please help me.** We already set `areas: false`, I think this is a reasonable precedent and we can set `joins: false` too; a default projection would be even better but it may be too late not to call that a bc break. Your thoughts? What behavior do you expect?

If we can't do that, then we definitely want to change our best practices around configuring the apostrophe-pages module for our own use. It would be nice to push them in the apostrophe cli, which might need a "strict mode" in which we promise not to do that and a "best practices" mode in which we actually try to not make your life painful.

**Comment by boutell** [comment]
* The fetching of all news items on the home page (and any page with a blog widget!) was due to a missing `criteria` argument in the implementation of `widgetCursor` in the `apostrophe-blog` and `apostrophe-events` modules. It got past us until now because of the algorithm that combines queries for all the widgets of the same type on the page and then pulls out the results each widget cares about; that was resulting in the right widgets on the page, but we were paying a huge performance price which, as more blog posts were added, was eventually going to bring sites to their knees. Events had the same bug. This is fixed, apostrophe-blog and apostrophe-events 2.0.1! Huge win.

* The other big win for UE was to do this:

```javascript
// in lib/modules/apostrophe-pages/index.js
  filters: {
    ancestors: {
      areas: false,
      joins: false,
      children: {
        areas: false,
        joins: false
      }
    },
    children: {
      areas: false,
      joins: false
    }
  },
```

Previously this site was using the default behavior, which is lousy and results in fetching every widget on the home page twice. I would like to change it, rather than having to document the above as a best practice boilerplate everybody has to use. Our intent to be conservative in fetching "related pages" has always been clear, but I don't know if that's enough for me to call this a bug fix rather than a bc break.

The UE home page is down to 33 queries, and I have an alpha-quality tool for logging all the cursors and a stack trace of where they came from. I'd like to get that into a module of its own rather than apostrophe core, it ought to be possible.

**Comment by boutell** [comment]
Further performance improvements are challenging. We should look at using the capacity of newer versions of MongoDB to carry out real joins.

This strategy might work as follows:

* Every time a doc is saved, all doc IDs inside it (i.e. joined to it) are located and added to a flat array, `joinedDocIds`.

* On page load, the query for the page includes a join based on this.

* Later, when widgets are loaded "by id," we first check and see if the join already got those. If it did, we simulate the projection as well as the application of any non-id criteria (an example: past events selected for an events widget must still get discarded based on their age), and return the winning objects without the need for a database query.

See the MongoDB documentation re: `$lookup` and `$unwind`:

https://docs.mongodb.com/manual/reference/operator/aggregation/lookup/#unwind-example

It is not yet known whether forcing every page query to be an aggregation query will have an unacceptable impact, that is, whether aggregation queries will break compatibility with other mongodb features we depend on.

But if they do, we could at least do *just one* extra query, for all of the joined docs. And that might make a good first step toward implementation of this approach anyway.

**Comment by boutell** [comment]
Work is underway to address this by implementing an optimization in which the ids of docs fetched as a result of joins from a doc on a clean, full fetch right after a save operation are recorded, and on a new request for the doc, those additional docs are automatically fetched immediately afterwards. The original query semantics are then faithfully implemented via the `sift` library. Where this cannot be done, the traditional code path is followed.

This should eliminate numerous unnecessary queries. In addition, it may be possible to use MongoDB's `$lookup` feature to implement this on the server side, saving another query.

**Comment by boutell** [comment]
@ludovicbret 

The 651 branch has a partial implementation of a query optimizer. Remarkably, this is eliminating close to 50% of the queries on the site I'm testing (a real production site with interesting content, although not one of yours so far).

So far overall performance isn't better when the site and mongodb are on the same dev machine. I'm looking for mistakes but it may simply be that the performance of running the queries through "sift," and also through my analyzer that makes sure it is safe to run them through "sift," is sufficiently inferior to MongoDB's C++ implementation that we lose the benefit of not sending the query to MongoDB and waiting for an answer.

But that is on the same machine. In your production environment, MongoDB and the webservers are much farther apart, and I hypothesize that this branch would be a big performance win in that environment.

I am continuing my work on this branch. Currently, it is too easy for the optimizer to become confused and believe it cannot eliminate a query that is actually a good candidate. To address this I will need to discover ids to be optimized in a slightly more thorough way.

**Comment by boutell** [comment]
See #1290 for the accommodations needed in core.
